File: swift/Tests/AmenClientTests/WebhookTests.swift
import XCTest
@testable import AmenClient

final class WebhookTests: XCTestCase {
    let secret = "unit-test-secret"

    // Golden vector from a real Amen delivery captured 2026-08-30.
    let goldenSecret = "whsec_test"
    let goldenTimestamp = "2026-08-30T18:53:23.885957+00:00"
    let goldenBody = Data(#"{"event":"deal.status.changed","timestamp":"2026-08-30T18:53:23.885957+00:00","payload":{"number":"D-0000000002","status":"paid"}}"#.utf8)
    let goldenHex = "950ca0ff7494dd435d4dc9d7e7ebe31cf54f0859a28a69a686d77e8db9dfd45c"

    func testGoldenVector() {
        let hex = WebhookSignature.compute(secret: goldenSecret, timestamp: goldenTimestamp, rawBody: goldenBody)
        XCTAssertEqual(hex, goldenHex)
        XCTAssertEqual(WebhookSignature.signatureHeaderValue(secret: goldenSecret, timestamp: goldenTimestamp, rawBody: goldenBody), "sha256=\(goldenHex)")
        XCTAssertTrue(WebhookSignature.verify(secret: goldenSecret, timestamp: goldenTimestamp, rawBody: goldenBody, received: "sha256=\(goldenHex)"))
    }

    func testSignatureRoundtrip() {
        let ts = "2026-08-30T18:53:23.885957+00:00"
        let body = Data(#"{"event":"deal.status.changed","timestamp":"2026-08-30T18:53:23.885957+00:00","payload":{}}"#.utf8)
        let sig = WebhookSignature.compute(secret: secret, timestamp: ts, rawBody: body)
        XCTAssertTrue(WebhookSignature.verify(secret: secret, timestamp: ts, rawBody: body, received: sig))
        XCTAssertTrue(WebhookSignature.verify(secret: secret, timestamp: ts, rawBody: body, received: "sha256=\(sig)"))
        // Tamper with the body.
        XCTAssertFalse(WebhookSignature.verify(secret: secret, timestamp: ts, rawBody: body + Data(" ".utf8), received: sig))
        // Tamper with the timestamp — it is part of the signed message.
        XCTAssertFalse(WebhookSignature.verify(secret: secret, timestamp: ts + "0", rawBody: body, received: sig))
        // Missing signature / missing timestamp.
        XCTAssertFalse(WebhookSignature.verify(secret: secret, timestamp: ts, rawBody: body, received: nil))
        XCTAssertFalse(WebhookSignature.verify(secret: secret, timestamp: nil, rawBody: body, received: sig))
    }

    func testHandlerVerifiesAndDedupes() async {
        let seen = Seen()
        let h = WebhookHandler(secret: secret) { e in await seen.add(e.id) }
        let ts = "2026-08-30T18:53:23.885957+00:00"
        let body = Data(#"{"event":"deal.status.changed","timestamp":"2026-08-30T18:53:23.885957+00:00","payload":{"number":"D-1","status":"paid"}}"#.utf8)
        func headers() -> [String: String] {
            ["X-Webhook-Signature": WebhookSignature.signatureHeaderValue(secret: secret, timestamp: ts, rawBody: body),
             "X-Webhook-Timestamp": ts,
             "X-Webhook-Event": "deal.status.changed"]
        }
        // Bad signature -> 401.
        let bad = await h.handle(headers: ["X-Webhook-Signature": "sha256=bad", "X-Webhook-Timestamp": ts], rawBody: body)
        XCTAssertEqual(bad.status, 401)
        // Missing timestamp header -> 401.
        let noTs = await h.handle(headers: ["X-Webhook-Signature": WebhookSignature.signatureHeaderValue(secret: secret, timestamp: ts, rawBody: body)], rawBody: body)
        XCTAssertEqual(noTs.status, 401)
        // First valid delivery -> 200.
        let first = await h.handle(headers: headers(), rawBody: body)
        XCTAssertEqual(first.status, 200)
        XCTAssertNil(first.body["duplicate"])
        // Duplicate (same timestamp) -> 200 + duplicate.
        let second = await h.handle(headers: headers(), rawBody: body)
        XCTAssertEqual(second.status, 200)
        XCTAssertEqual(second.body["duplicate"] as? Bool, true)
        // De-dupe keyed on the event timestamp.
        let ids = await seen.ids
        XCTAssertEqual(ids, [ts])
    }
}
actor Seen { var ids: [String] = []; func add(_ id: String) { ids.append(id) } }

File: csharp/tests/Amen.Kit.Tests/WebhookTests.cs
using System.Text;
using Amen.Kit.Webhooks;
using Xunit;

public class WebhookTests
{
    // Golden vector captured from a real Amen delivery (2026-08-30).
    const string GoldenSecret = "whsec_test";
    const string GoldenTimestamp = "2026-08-30T18:53:23.885957+00:00";
    const string GoldenBody = "{\"event\":\"deal.status.changed\",\"timestamp\":\"2026-08-30T18:53:23.885957+00:00\",\"payload\":{\"number\":\"D-0000000002\",\"status\":\"paid\"}}";
    const string GoldenHex = "950ca0ff7494dd435d4dc9d7e7ebe31cf54f0859a28a69a686d77e8db9dfd45c";

    static byte[] Body => Encoding.UTF8.GetBytes(GoldenBody);

    [Fact] public void ComputeMatchesGoldenVector()
    {
        Assert.Equal(GoldenHex, WebhookSignature.Compute(GoldenSecret, GoldenTimestamp, Body));
    }

    [Fact] public void VerifyAcceptsBareAndPrefixedSignatures()
    {
        Assert.True(WebhookSignature.Verify(GoldenSecret, GoldenTimestamp, Body, GoldenHex));
        Assert.True(WebhookSignature.Verify(GoldenSecret, GoldenTimestamp, Body, "sha256=" + GoldenHex));
        Assert.True(WebhookSignature.Verify(GoldenSecret, GoldenTimestamp, Body, "sha256=" + GoldenHex.ToUpperInvariant()));
    }

    [Fact] public void VerifyRejectsTamperedBodyTimestampAndSecret()
    {
        // Tampered body.
        Assert.False(WebhookSignature.Verify(GoldenSecret, GoldenTimestamp, Encoding.UTF8.GetBytes(GoldenBody + " "), GoldenHex));
        // Wrong timestamp (part of the signed message).
        Assert.False(WebhookSignature.Verify(GoldenSecret, "2026-08-30T18:53:23.000000+00:00", Body, GoldenHex));
        // Wrong secret.
        Assert.False(WebhookSignature.Verify("whsec_other", GoldenTimestamp, Body, GoldenHex));
    }

    [Fact] public void VerifyRejectsMissingSignatureOrTimestamp()
    {
        Assert.False(WebhookSignature.Verify(GoldenSecret, GoldenTimestamp, Body, null));
        Assert.False(WebhookSignature.Verify(GoldenSecret, GoldenTimestamp, Body, ""));
        Assert.False(WebhookSignature.Verify(GoldenSecret, null, Body, GoldenHex));
        Assert.False(WebhookSignature.Verify(GoldenSecret, "", Body, GoldenHex));
    }

    static IEnumerable<KeyValuePair<string, string>> Headers(string? sig, string? ts, string? evt = "deal.status.changed")
    {
        var h = new List<KeyValuePair<string, string>>();
        if (sig is not null) h.Add(new(WebhookSignature.SignatureHeader, sig));
        if (ts is not null) h.Add(new(WebhookSignature.TimestampHeader, ts));
        if (evt is not null) h.Add(new(WebhookSignature.EventHeader, evt));
        return h;
    }

    [Fact] public async Task HandlerRejectsBadSignatureWith401()
    {
        var h = new WebhookHandler(GoldenSecret, _ => Task.CompletedTask);
        Assert.Equal(401, (await h.HandleAsync(Headers("sha256=deadbeef", GoldenTimestamp), Body)).Status);
        // Missing timestamp header also fails verification.
        Assert.Equal(401, (await h.HandleAsync(Headers("sha256=" + GoldenHex, null), Body)).Status);
    }

    [Fact] public async Task HandlerVerifiesDispatchesAndDedupesOnTimestamp()
    {
        var seen = new List<WebhookEvent>();
        var h = new WebhookHandler(GoldenSecret, e => { seen.Add(e); return Task.CompletedTask; });
        var good = Headers("sha256=" + GoldenHex, GoldenTimestamp);

        var first = await h.HandleAsync(good, Body);
        Assert.Equal(200, first.Status);
        Assert.True((bool)first.Body["ok"]);

        var second = await h.HandleAsync(good, Body);
        Assert.Equal(200, second.Status);
        Assert.Equal(true, second.Body["duplicate"]);

        Assert.Single(seen);
        Assert.Equal("deal.status.changed", seen[0].Type);
        Assert.Equal(GoldenTimestamp, seen[0].Id);
    }
}
